crates/valence_network/src/legacy_ping.rs

```diff
@@ -0,0 +1,305 @@
+        // effectively insignificant too. Network IO is just not that reliable
+        // at this level, the connection may be lost as well or something at this point.
+        sleep(Duration::from_millis(10)).await;
+        n = stream.peek(&mut temp_buf).await?;
+    }
+
+    let format = match &temp_buf[..n] {
+        [0xfe] => PingFormat::Pre1_4,
+        [0xfe, 0x01] => PingFormat::Pre1_6,
+        [0xfe, 0x01, 0xfa] => PingFormat::Pre1_7,
+        _ => return Ok(false), // Not a legacy ping
+    };
+
+    let payload = match format {
+        PingFormat::Pre1_7 => Some(read_payload(stream).await?),
+        _ => None,
+    };
+
+    if let ServerListLegacyPing::Respond(response) = shared
+        .0
+        .callbacks
+        .inner
+        .server_list_legacy_ping(shared, remote_addr, payload)
+        .await
+    {
+        let separator = match format {
+            PingFormat::Pre1_4 => '§',
+            _ => '\0',
+        };
+
+        let mut buf = BytesMut::new();
+
+        // packet ID and length placeholder
+        buf.extend([0xff, 0x00, 0x00]);
+
+        if format != PingFormat::Pre1_4 {
+            // some constant bytes lol
+            buf.extend("§1\0".encode_utf16().flat_map(|c| c.to_be_bytes()));
+
+            // protocol and version
+            buf.extend(
+                format!(
+                    "{protocol}{separator}{version}{separator}",
+                    protocol = response.protocol,
+                    version = response.version
+                )
+                .encode_utf16()
+                .flat_map(|c| c.to_be_bytes()),
+            );
+        }
+
+        buf.extend(
+            if format == PingFormat::Pre1_4 {
+                // remove formatting for pre-1.4 legacy pings
+                remove_formatting(&response.description)
+            } else {
+                response.description
+            }
+            .encode_utf16()
+            .flat_map(|c| c.to_be_bytes()),
+        );
+
+        // Online and max players
+        buf.extend(
+            format!(
+                "{separator}{online_players}{separator}{max_players}",
+                online_players = response.online_players,
+                max_players = response.max_players
+            )
+            .encode_utf16()
+            .flat_map(|c| c.to_be_bytes()),
+        );
+
+        // replace the length placeholder with the actual length
+        let chars = (buf.len() as u16 - 3) / 2; // -3 because of the packet prefix (id and length), and /2 because UTF16
+        buf[1..3].copy_from_slice(chars.to_be_bytes().as_slice());
+
+        stream.write_all(&buf).await?;
+    }
+
+    Ok(true)
+}
+
+async fn read_payload(stream: &mut TcpStream) -> io::Result<ServerListLegacyPingPayload> {
+    // consume the first 29 useless bytes of this amazing protocol
+    stream.read_exact(&mut [0u8; 29]).await?;
+
+    let protocol = stream.read_u8().await? as i32;
+    let hostname_len = stream.read_u16().await? * 2;
```

Comment:
This multiplication could overflow and panic in debug mode. Also might want to think about bounding this to some reasonable limit.